### ISSUE TITLE
Client support for handling cookies

### ISSUE THREAD

**Issue by Ubehebe** [line/armeria#2637]
First, thank you for your work on Armeria. I've been using the framework since v0.84 to host a mix of HTTP and gRPC services. I'm consistently impressed with the quality of Armeria's APIs and documentation, and the energy and positive attitude you bring to the open-source community.

This is the first time I've had a problem with the framework that I need help to solve. I have a simple HTTP endpoint (implemented as an annotated service) that works like this:

```
GET / -> You have visited 0 times
GET / -> You have visited 1 times
GET / -> You have visited 2 times
GET / -> You have visited 3 times
GET / -> You have visited 4 times
...
```

The endpoint tracks unique visits by setting a cookie on the first visit and keeping an in-memory map of how many times it's seen that cookie.

The client in the above pseudocode can be any piece of software that understands how cookies work. I've written a version of the test that uses curl (via its `--cookie` and `--cookie-jar` flags), and a version that brings up a full browser. When tested 100 times, both versions pass 100% of the time. So I believe my server-side code is correct.

I would like to write a version of the test that uses Armeria's WebClient. When I wrote such a test (using a minimal `WebClient.of(url)`) I observed the following:

```
GET / -> You have visited 0 times
GET / -> You have visited 0 times
GET / -> You have visited 0 times
GET / -> You have visited 0 times
GET / -> You have visited 0 times
...
```

The server sent the `Set-Cookie` header on every response, but the Armeria client never set the corresponding `Cookie` header on requests.

OK, so WebClient doesn't do anything with cookies by default. I didn't see any option in WebClientBuilder or related classes for handling cookies.

My next thought was: I could use a decorator to wrap a WebClient, intercept the response, read the `Set-Cookie` header into an in-memory data structure ("cookie jar"), then, before the next request goes out, set the `Cookie` header from the cookie jar. This was my first effort:

```java
private static Function<HttpClient, HttpClient> decorateWithCookieJar(CookieJar cookieJar) {
  return underlying ->
      new HttpClient() {
        @Override
        public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
          // in browsers, an origin is defined as scheme + hostname + port. This should be close
          // enough for tests.
          var origin = req.scheme() + req.authority();
          cookieJar.get(origin).forEach(cookie -> ctx.addAdditionalRequestHeader(COOKIE, cookie));
          var response = underlying.execute(ctx, req).toDuplicator();
          response
              .duplicate()
              .aggregate()
              .thenApply(r -> r.headers().getAll(SET_COOKIE))
              .thenAccept(cookies -> cookieJar.putAll(origin, cookies));
          return response.duplicate();
        }
      };
}
```

(`CookieJar` is a small class that holds a map of cookies for each origin.) This seemed to work at first. However, when I ran the above test 100 times against the Armeria client, it failed about 30% of the time, like this:

```
GET / -> You have visited 0 times
GET / -> You have visited 0 times // FAIL: should be 1
GET / -> You have visited 1 times // FAIL: should be 2
...
```

After some debugging, I think I understand the bug here. The decorator duplicates the `HttpResponse`, using one copy to write the cookies into the cookie jar, and the other copy to return to the caller. (It doesn't seem possible to use a single `HttpResponse` for both tasks, though please correct me if I'm wrong.) But these copies aren't synchronized. If the copy used for populating the cookie jar doesn't resolve until after the next call to WebClient.get(), the cookie jar won't contain the cookie, leading to the observed test failure.

I tried a couple of other approaches, including using the RequestLog API:
```java
ctx.log().whenComplete().thenAccept(
    requestLog -> /* populate cookie jar from request log */);
```

But the test was still about 30% flaky.

The fundamental issue is my lack of familiarity with Armeria's concurrency model. Many Armeria APIs take EventExecutor arguments, including the toDuplicator() and aggregate() calls in the code above. But Armeria has high-level APIs for so many use cases that dropping down to this level feels scary. I would like an API to express my intent: that the WebClient instance must have finished processing the previous response before it fires off the next request.

I think this issue could be resolved in several ways. Here are some possible resolutions, ordered by what seems to me increasing benefit to Armeria users:

1. My problem could be solved by configuring the decorator using an existing API. (Is there an API to say "make this WebClient single-threaded"? That would seem to solve this problem.)
2. My problem could be solved by improving the documentation of Armeria's concurrency model. A document like "How to do X in Armeria without blocking/deadlocking" would be helpful. Issues like #2064 suggest this is a pain point in the community.
3. My problem could be solved by making WebClient (or really any HttpClient) understand cookies out of the box. This seems like a common enough use case, and JDK 11's HttpClient supports cookies. Most importantly, I could maintain my ignorance of Armeria's concurrency model :) Hypothetical APIs:
```java
var webClient = WebClient.builder().useDefaultCookieJar().build();
var webClient = WebClient.builder().useCookieJar(/* allow callers to pass in a cookie jar so they can inspect it later */).build();
```

Thanks again for all of your work.

**Comment by anuraaga** [comment]
Thanks for the suggestion. Having a simple API like a prebaked decorator or `ClientOption` for handling cookies seems like it would be convenient.

In the meantime, I think you can solve your issue by using `FilteredHttpResponse`.

https://github.com/line/armeria/blob/577f7a73dc4f3c1c6f315ca1bd3da33a0de7d193/core/src/main/java/com/linecorp/armeria/common/FilteredHttpResponse.java

This would make it very explicit that you process headers as they come in.

Another approach with less code which I think would work is using the request log and use `whenAvailable(RequestLogProperty.RESPONSE_HEADERS`. It less explicitly has your code intercept the logic but should still be guaranteed to run before the response completes whereas `whenComplete` runs after.

**Comment by Ubehebe** [comment]
Thanks for the quick response. Confirmed that the `whenAvailable(RESPONSE_HEADERS).thenAccept(...)` approach eliminates all test flakiness. I'll try the FilteredHttpResponse approach tomorrow as well.

> It less explicitly has your code intercept the logic but should still be guaranteed to run before the response completes whereas whenComplete runs after.

This is exactly the property I was looking for. However, I don't think I could have figured this out just by reading the code/documentation/javadocs.

**Comment by trustin** [comment]
@Ubehebe wrote:

```
var webClient = WebClient.builder().useDefaultCookieJar().build();
var webClient = WebClient.builder().useCookieJar(/* allow callers to pass in a cookie jar so they can inspect it later */).build();
```

This sounds good to me. It will be a great addition/shortcut for Armeria clients.

And thanks a lot for your kind words! :heart: We look forward to getting more feed back from you. :bowing_man:

**Comment by Ubehebe** [comment]
Before I start implementing this, it's probably a good idea to have a discussion about requirements and design.

My initial thought had been to write a simple CookieJar class that callers could pass in while building the WebClient, as in the pseudocode above. CookieJar would be an abstraction over a simple map of origins to cookies.

The problem with this approach is that the cookie-matching rules ("should I send cookie X for URI Y?") are not trivial. They are specified by [RFC 6265](https://tools.ietf.org/html/rfc6265) and involve things like path and domain matching. Reimplementing that logic would be an invitation for bugs.

The JDK has shipped with a few cookie classes in java.net since Java 6, in particular [CookieStore](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/CookieStore.html) for storage and [CookieHandler](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/CookieHandler.html) adding the RFC semantics. Would you be open to using them in public Armeria APIs? Example:

```java
WebClientBuilder.cookieHandler(java.net.CookieHandler cookieHandler);
```

This might involve some translation back and forth between the Armeria cookie representation and the JDK's (java.net.HttpCookie). The JDK cookie APIs are not as nice as Armeria's (more mutable state, weirder data structures), but the value of this is not having to reimplement the RFC cookie semantics.

Thoughts?

**Comment by trustin** [comment]
How about designing a nice API and then wrapping JDK's implementation?

**Comment by ikhoon** [comment]
Fixed by #3118
